[somfytahoma] Migration to the bnd based build system

octa22 opened the pull request:
see #5005

automigrated using migrate2bnd.bash

Signed-off-by: Ondrej Pecta <opecta@gmail.com>

octa22 commented:
Hi guys,

what I found weird is autogenerated MANIFET.MF import 
`com.google.gson;version="[2.8,3)"`

which generates
`Unresolved requirement: Import-Package: com.google.gson; version="[2.8.0,3.0.0)"` 
in OH 2.4
Shouldn't we use rather
`com.google.gson;version="[2.3,3)"`
Thanks

davidgraeff commented:
Everything on the master branch is for the next release. Oh 2.4 was a snapshot of master at that time and is tagged accordingly.

davidgraeff commented on bundles/org.openhab.binding.somfytahoma/src/main/java/org/openhab/binding/somfytahoma/internal/handler/SomfyTahomaBridgeHandler.java:
```diff
@@ -61,18 +62,20 @@
     /**
      * Future to poll for updates
      */
+    @Nullable
```
Please use `private @Nullable ...` instead

davidgraeff commented on bundles/org.openhab.binding.somfytahoma/src/main/java/org/openhab/binding/somfytahoma/internal/SomfyTahomaHandlerFactory.java:
```diff
@@ -157,16 +159,14 @@ protected ThingHandler createHandler(Thing thing) {
-                discoveryServiceRegs.remove(thingHandler.getThing().getUID());
+            // remove discovery service, if bridge handler is removed
+            SomfyTahomaItemDiscoveryService service = (SomfyTahomaItemDiscoveryService) bundleContext
+                    .getService(serviceReg.getReference());
```
This is the old way of having a thing bound service.

If you have a minute, could you adapt this to the new way? (A thing handler nowadays has a `getServices()` method. Overwrite it, return `SomfyTahomaItemDiscoveryService.class`, implement the suggested interface. Thanks.

octa22 commented:
> Nice code improvements. I have left two inline comments.
> I tend to merge this PR without those changes and hopping for a follow up PR. WDYT?

OK, you can merge it, the next PR with these changes will follow. thanks

octa22 commented:
Hi @davidgraeff ,
tried to use the way of registering devices but no luck. What I did:
1) Overridden `getServices` in bridge handler, returning `SomfyTahomaItemDiscoveryService.class`
2) Implemented interface `ThingHandlerService` in `SomfyTahomaItemDiscoveryService` (including defaut contructor, `getThingHandler` & `setThingHandler` methods)
3) removed methods `registerItemDiscoveryService` and `removeHandler` from `SomfyTahomaHandlerFactory` class

The `SomfyTahomaItemDiscoveryService` gets registered automatically (constructor executed) but background discovery service is not started, in PaperUI there is no possibility of manual discovery execution for Somfy binding. (Somfy binding not in a list)
Any idea what I am missing? Shoud I register the discovery service in `SomfyTahomaHandlerFactory` class and call `activate` method? 
Thanks.

davidgraeff commented:
> gets registered automatically (constructor executed)

Perfect. You must annotate your class as `@Component(service=DiscoveryServiceForgotTheInterfaceName.class)`. You obviously use the discovery service interface name, I'm too lazy to look that up. openHAB will register discovery services by listening for OSGi components that announce this interface.

cweitkamp commented:
@davidgraeff I seem to have missed something. Did you find a way how to get the discovery to work using this way? IIRC we agreed on a bug in the core framework for registering discovery services in #4921 .

octa22 commented:
Hi,

I am really convinced there must be some bug or I am doing something wrong.

1) the `getServices` approach instantiates discovery service and do not starts background discovery (seems to be registered not as a `DiscoveryService`)

2) the `@component(service...)` approach instantiates discovery service including background discovery, but the bridge handler is null (so the `setThingHandler` is not called)

3) the combination of the two methods results in two instances of `SomfyTahomItemDiscoveryService`. None of them is working OK.

Thanks

Ondrej

davidgraeff commented:
Hm. The component should not start immediately and if it does, immediate=false must be set as we really don't want OSGi to start this component.

@cweitkamp This is still the way how to register thing services although the core has that wrong condition check. 

People over at core are super careful with merging anything apart from buildsystem changes is my experience. But if we build up a bit of pressure we can fix that core issue this week, which was really a one liner if I recall correctly

cweitkamp commented:
Our first assumption was that the interface name matching does not work. This is not yet invalidated. I tried to tackle it down and currently I have the following guess: The [ThingHandlerFactory](https://github.com/openhab/openhab-core/blob/master/bundles/org.openhab.core.thing/src/main/java/org/eclipse/smarthome/core/thing/binding/BaseThingHandlerFactory.java#L147) calls `activate()` on the `ThingHandlerService`. And thus it does not call the `activate()` method of the `AbstractDiscoveryService`. A solution could be to simply check if we have a `DiscoveryService` instance and cast it accordingly before activation.

cweitkamp commented:
> we can fix that core issue this week. It is really a one liner if I recall correctly

Yes, you did. And I am pretty sure we can easily cover it by a unit / integration test.

davidgraeff commented:
True. I think someone made it work in his new binding PR by implementing activate and deactivate which speaks for your theory. A cast might help indeed.

cweitkamp commented:
I will try to come up with an OHC PR in the next couple of days to make it work.
